### PR TITLE
Issue Addressing

### PR DESCRIPTION
**Pull request by aspriddell**
This merge covers small parts of issues and some things I've added myself:

- I've added a RS Beatmap Struct with JSON Headings for decoding the maps with minimal code. It's just the class at the moment with no functions - they'll need to be implemented later (and will probably then be converted/cast to the Beatmap class) (#32)
- When you press space the track `mainmenu_aqours` will start playing
- If the side drawer is open, the escape key will close it
- The menu button at the top right corner has been removed and the same functionality resides in clicking the player's icon (less space needed and can be used somewhere else)
- The `osu.Framework` has been updated a but there are some breaking changes I haven't fixed yet.
- Added difficulty enum (to compliment RS formatted maps)
- Fixed Currency Plus buttons, so they show the `NewValue` instead of the class name

### CODE REVIEW

File: Lovewing/Beatmaps/MapDifficulties.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Lovewing.Vars
```
Comment (jorolf) [inline Lovewing/Beatmaps/MapDifficulties.cs]: Please use the correct namespace

File: Lovewing/Beatmaps/RS/RSFormat.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public class RSFormat
+    {
+        [JsonProperty("song_name")]
+        public string SongName { get; set; }
+
+        [JsonProperty("difficulty")]
+        public int Difficulty { get; set; }
+
+        [JsonProperty("rank_info")]
+        public IEnumerable<RSRankInfo> Ranks { get; set; }
```
Comment (jorolf) [inline Lovewing/Beatmaps/RS/RSFormat.cs]: wouldn't it be better to just use an array or a list directly?

File: Lovewing/Graphics/Containers/TabContainer.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
 using System;
+using SpriteIcon = Lovewing.Graphics.Sprites.SpriteIcon;
```
Comment (jorolf) [inline Lovewing/Graphics/Containers/TabContainer.cs]: if you're going to update the framework version then also use the framework's ``SpriteIcon`` and ``FontAwesome``

File: Lovewing/Graphics/UserInterface/LovewingTextBox.cs [hunk 4]
```diff
@@ -15,9 +15,9 @@ public class LovewingTextBox : TextBox
 
         protected override float LeftRightPadding => 10;
 
-        protected override Color4 BackgroundCommit => CommitColour;
-        protected override Color4 BackgroundFocused => FocusedColour;
-        protected override Color4 BackgroundUnfocused => UnfocusedColour;
+        //protected override Color4 BackgroundCommit => CommitColour;
+        //protected override Color4 BackgroundFocused => FocusedColour;
+        //protected override Color4 BackgroundUnfocused => UnfocusedColour;
```
Comment (jorolf) [inline Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: I'd remove them completely and not just comment them out. Although I'm not quite sure why you commented them out in the first place....

File: Lovewing/Lovewing.csproj [hunk 5]
```diff
@@ -16,7 +16,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="ppy.osu.Framework" Version="2019.313.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
```
Comment (jorolf) [inline Lovewing/Lovewing.csproj]: ppy.osu.Framework already depends on Newtonsoft.Json so it should already be imported

File: Lovewing/Screens/Game/SongSelectorScreen.cs [hunk 6]
```diff
@@ -156,7 +157,7 @@ protected override void LoadComplete()
                             Origin = Anchor.TopCentre,
                             Margin = new MarginPadding { Top = 20 },
                             Text = selected.Value.Title,
-                            TextSize = 40,
+                            Height = 40,
```
Comment (jorolf) [inline Lovewing/Screens/Game/SongSelectorScreen.cs]: That's not how you change the textsize

File: Lovewing/Screens/MainScreen.cs [hunk 7]
```diff
@@ -24,17 +32,26 @@ public class MainScreen : Screen
         private TabContainer liveshowTab;
         private GameScreen gameScreen;
         private SongSelectorScreen songSelectorScreen;
+        private TextureStore textureStore;
+        private Track Theme;
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: naming

File: Lovewing/Screens/MainScreen.cs [hunk 8]
```diff
@@ -172,6 +189,24 @@ private void selectTab(VisibilityContainer container, Visibility visibility)
...
+            {
+                case osuTK.Input.Key.Space:
+                    if (Theme.IsRunning)
+                        Theme.Stop();
+                    else
+                        Theme.Start();
+                    break;
+                case osuTK.Input.Key.Escape:
+                    if (((LovewingGame)Game).Sidebar.IsOpen)
+                        ((LovewingGame)Game).Sidebar.Toggle();
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: Isn't ``Hide()`` and thing?

File: Lovewing/Screens/MainScreen.cs [hunk 9]
```diff
@@ -24,17 +32,26 @@ public class MainScreen : Screen
...
+        {
+            Theme = audio.Track.Get("mainmenu_aqours");
+            textureStore = textures;
+        }
+
         protected override bool OnClick(ClickEvent state)
         {
-            if (((LovewingGame) Game).Sidebar.IsOpen)
-                ((LovewingGame) Game).Sidebar.Toggle();
+            if (((LovewingGame)Game).Sidebar.IsOpen)
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: Add a line like this somewhere:

https://github.com/ppy/osu/blob/98626018fdd0f06d5e3185d6831fc09e35a6d0cb/osu.Game/Screens/OsuScreen.cs#L52-L54

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (aspriddell) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: I think that each song has a different score requirement for each rank, so the "ranks" are a number and the max_rank is something to do with the min. score to achieve that rank. As far as I know this is correct as the maps' ranks all differentiate.

File: Lovewing/Graphics/Containers/TabContainer.cs (hunk 3 above)
Comment (aspriddell) [reply Lovewing/Graphics/Containers/TabContainer.cs]: tried now - it causes issues with downstream code and the ButtonIcon field does not work as it is a static class. Using the Lovewing one works fine with no issues

File: Lovewing/Graphics/UserInterface/LovewingTextBox.cs (hunk 4 above)
Comment (aspriddell) [reply Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: they're incompatible with the newer version.

File: Lovewing/Screens/Game/SongSelectorScreen.cs (hunk 6 above)
Comment (aspriddell) [reply Lovewing/Screens/Game/SongSelectorScreen.cs]: I'll revert this one but for future reference that method was made obsolete...

File: Lovewing/Screens/MainScreen.cs (hunk 8 above)
Comment (aspriddell) [reply Lovewing/Screens/MainScreen.cs]: yes but for this I have changed it to `ToggleVisibility()` Instead and changed the key to tab for easier navigation

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (jorolf) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: I meant the type. Instead of ``IEnumerable`` you could use and array or a list

File: Lovewing/Graphics/Containers/TabContainer.cs (hunk 3 above)
Comment (jorolf) [reply Lovewing/Graphics/Containers/TabContainer.cs]: I still think that you should fix the problems and use them instead

File: Lovewing/Screens/Game/SongSelectorScreen.cs (hunk 6 above)
Comment (jorolf) [reply Lovewing/Screens/Game/SongSelectorScreen.cs]: Yes, use the new version then instead of the deprecated one

File: Lovewing/Screens/MainScreen.cs [hunk 10]
```diff
@@ -13,29 +13,47 @@
...
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Input.Bindings;
+using Lovewing.KeyBindings.MainMenu;
+using osu.Framework.Audio.Track;
 
 namespace Lovewing.Screens
 {
     public class MainScreen : Screen
     {
+        protected new LovewingGame Game => base.Game as LovewingGame;
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: Use a direct cast and not ``as``, Game should always be ``LovewingGame`` anyway

File: Lovewing/Screens/MainScreen.cs [hunk 11]
```diff
@@ -172,6 +190,23 @@ private void selectTab(VisibilityContainer container, Visibility visibility)
...
+            switch(e.Key)
+            {
+                case osuTK.Input.Key.Space:
+                    if (theme.IsRunning)
+                        theme.Stop();
+                    else
+                        theme.Start();
+                    break;
+                case osuTK.Input.Key.Tab:
+                    ((LovewingGame)Game).Sidebar.ToggleVisibility();
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: unnecessary now
```suggestion
                    Game.Sidebar.ToggleVisibility();
```

File: Lovewing/Graphics/UserInterface/LovewingTextBox.cs (hunk 4 above)
Comment (jorolf) [reply Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: Now that I think about it, you probably want to inherit from ``BasicTextBox`` instead of ``TextBox`` for now

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (aspriddell) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: having it as an `IEnumerable` would make it easier to perform LINQ operations on it and can be converted to a List, `IOrderedEnumerable` or any form of list you wish

File: Lovewing/Graphics/Containers/TabContainer.cs (hunk 3 above)
Comment (aspriddell) [reply Lovewing/Graphics/Containers/TabContainer.cs]: sorted - you can now use the `osu.Framework.Graphics.Sprites.FontAwesome` and have an even larger array of icons

File: Lovewing/Screens/Game/SongSelectorScreen.cs (hunk 6 above)
Comment (aspriddell) [reply Lovewing/Screens/Game/SongSelectorScreen.cs]: sorted - and have fixed all the other depreciation warnings

File: Lovewing/Screens/MainScreen.cs [hunk 12]
```diff
@@ -13,29 +13,46 @@
...
         {
-            if (((LovewingGame) Game).Sidebar.IsOpen)
-                ((LovewingGame) Game).Sidebar.Toggle();
+            theme = audio.Track.Get("mainmenu_aqours");
+            textureStore = textures;
+        }
 
+        protected override bool OnClick(ClickEvent state)
+        {
+            Game.Sidebar.Hide();
```
Comment (aspriddell) [inline Lovewing/Screens/MainScreen.cs]: doesn't need if statement and works fine as is. The if statement would cause a block where it wouldn't hide (by clicking) until you pressed tab (to toggle) then reopen it (by clicking)

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (jorolf) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: You can also use LINQ with arrays or lists because they both implement ``IEnumerable``

Comment (aspriddell) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: true but line operations return an `IEnumerable<T>`, so would have to suffix `.ToList()` to each operation at the end

Comment (jorolf) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: How often would this be set and how often would it be read? I'd think that an array or a list has a bigger advantage here because you can access and set any member at any time.

CC: @sr229

File: Lovewing/Graphics/Containers/TabContainer.cs [hunk 13]
```diff
@@ -7,6 +7,7 @@
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
 using System;
+using SpriteIcon = osu.Framework.Graphics.Sprites.SpriteIcon;
```
Comment (jorolf) [inline Lovewing/Graphics/Containers/TabContainer.cs]: Not needed anymore?

File: Lovewing/Graphics/Overlays/Toolbar.cs [hunk 14]
```diff
@@ -9,6 +9,7 @@
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Graphics.Textures;
 using System;
+using SpriteIcon = Lovewing.Graphics.Sprites.SpriteIcon;
```
Comment (jorolf) [inline Lovewing/Graphics/Overlays/Toolbar.cs]: use osu.Framework.Graphics.Sprites.SpriteIcon instead

File: Lovewing/Graphics/UserInterface/LovewingSmallButton.cs [hunk 15]
```diff
@@ -5,6 +5,7 @@
 using osu.Framework.Graphics.Colour;
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Graphics.Sprites;
+using SpriteIcon = Lovewing.Graphics.Sprites.SpriteIcon;
```
Comment (jorolf) [inline Lovewing/Graphics/UserInterface/LovewingSmallButton.cs]: ditto

File: Lovewing/Graphics/UserInterface/LovewingSmallButton.cs [hunk 16]
```diff
@@ -23,7 +24,7 @@ public FontAwesome Icon
         public string Font
```
Comment (jorolf) [inline Lovewing/Graphics/UserInterface/LovewingSmallButton.cs]: Use ``FontUsage`` instead of ``Font``.

Now that I think about it, why do we need to change the Font anyway...

File: Lovewing/Screens/Game/SongSelectorScreen.cs [hunk 17]
```diff
@@ -26,6 +27,7 @@ public class SongSelectorScreen : Screen
         private SpriteText titleText;
         private SpriteText artistText;
         private ScrollContainer beatmapContainer;
+        private FontUsage defaultFont = new FontUsage().With(size: 40);
```
Comment (jorolf) [inline Lovewing/Screens/Game/SongSelectorScreen.cs]: Do we really need to create a field for this?

File: Lovewing/Screens/MainScreen.cs [hunk 18]
```diff
@@ -13,29 +13,46 @@
 using osu.Framework.Screens;
 using System;
 using osu.Framework.Input.Events;
+using FontAwesome = Lovewing.Graphics.FontAwesome;
+using SpriteIcon = Lovewing.Graphics.Sprites.SpriteIcon;
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: ditto

File: Lovewing/Screens/MainScreen.cs [hunk 19]
```diff
@@ -55,7 +72,7 @@ protected override void LoadComplete()
                         {
                             Margin = new MarginPadding { Right = 100 },
                             ButtonText = @"Live",
-                            ButtonIcon = FontAwesome.fa_music,
+                            ButtonIcon = osu.Framework.Graphics.Sprites.FontAwesome.Solid.Music,
```
Comment (jorolf) [inline Lovewing/Screens/MainScreen.cs]: ditto

File: Lovewing/Graphics/UserInterface/LovewingTextBox.cs (hunk 4 above)
Comment (sr229) [reply Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: This is not wise to do it this way, you should port this instead since it might break other parts of the program

Comment (sr229) [reply Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: This is not wise to do it this way, you should port this instead since it might break other parts of the program

File: Lovewing/Graphics/Containers/TabContainer.cs (hunk 13 above)
Comment (sr229) [reply Lovewing/Graphics/Containers/TabContainer.cs]: this is way too unclean. Imports should be standard imports.

File: Lovewing/Beatmaps/MapDifficulties.cs [hunk 20]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Lovewing.Beatmaps
+{
+    public enum Difficulties
```
Comment (sr229) [inline Lovewing/Beatmaps/MapDifficulties.cs]: I don't think this is standard diffset for SIF. Can anyone verify?

CC @MikuAuahDark

File: Lovewing/Beatmaps/Note.cs [hunk 21]
```diff
@@ -1,12 +1,23 @@
-﻿namespace Lovewing.Beatmaps
+﻿using Newtonsoft.Json;
+
+namespace Lovewing.Beatmaps
 {
     public struct Note
     {
+        [JsonProperty("timing_sec")]
```
Comment (sr229) [inline Lovewing/Beatmaps/Note.cs]: This is gonna be hard to port when we decide to suppport other formats like LS2's NBT-based beatmap

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (sr229) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: We follow osu! paradigm for this, so I don't think this is correct.

**Comment by sr229** [comment]
I also want to add the Main Menu OST plays all the time, it cannot be paused.

File: Lovewing/Beatmaps/MapDifficulties.cs (hunk 20 above)
Comment (aspriddell) [reply Lovewing/Beatmaps/MapDifficulties.cs]: https://github.com/ProjectLovewing/Lovewing.Resources/tree/master/Lovewing.Game.Resources/Beatmaps/rs

**Comment by aspriddell** [comment]
> I also want to add the Main Menu OST plays all the time, it cannot be paused.

With or without looping?

**Comment by sr229** [comment]
The soundtrack loops infinitely so yes.

**Comment by aspriddell** [comment]
ok that can be done, is there a preference which one plays at the moment?

**Comment by sr229** [comment]
For now we haven't implemented the logic for that, OSTs are bound to your center girl's choice, whatever group your center girl will play the respective OST, if its a N class idol, default to Muse

**Comment by aspriddell** [comment]
Ok so just play and loop any for now.

File: Lovewing/Graphics/UserInterface/LovewingTextBox.cs (hunk 4 above)
Comment (aspriddell) [reply Lovewing/Graphics/UserInterface/LovewingTextBox.cs]: [osu Framework Update](https://github.com/ppy/osu-framework/blob/5e3fef4f8564742a2ecf808c88859c34cde11946/osu.Framework/Graphics/UserInterface/TextBox.cs#L72) meant that the override won't work, so I have just set them in the constructor for now. A BasicTextBox already has the background set in [the constructor](https://github.com/ppy/osu-framework/blob/5e3fef4f8564742a2ecf808c88859c34cde11946/osu.Framework/Graphics/UserInterface/BasicTextBox.cs#L18)

**Comment by aspriddell** [comment]
this now plays the ost at the start and is set to loop, all the icons have been set to the osu framework and as a result, removed the lovewing fontawesome class

**Comment by aspriddell** [comment]
I started to change the `Note` class to an interface and then all the other map formats can then implement it, I will end up doing the same to the `Beatmap` class eventually. This should fix the concern of adding new formats easily.

File: Lovewing/Beatmaps/MapDifficulties.cs (hunk 20 above)
Comment (MikuAuahDark) [reply Lovewing/Beatmaps/MapDifficulties.cs]: Sorry for the long reply, I was in college at the time I received the notification via E-mail and I wasn't logged in to GitHub in my phone.

I don't think that was correct difficulty enumeration. [Here's](https://www.dropbox.com/sh/e6rknlkhus8z8kb/AABeHWxhodbz16cjhoBpcO2da?dl=0&preview=SIFTrain+Extended+Beatmap+Legend.txt) the current SIFTrain beatmap format specification. Notice there's additional "Master" enum there. Also note that some fields are not used/supported in SIFTrain but it's used in livesim2, and such fields are marked with asterisk

> Denotes data members that were added to the Extended format which used by Live Simulator: 2 and are not used on the SIFTrain app

File: Lovewing/Beatmaps/Note.cs (hunk 21 above)
Comment (MikuAuahDark) [reply Lovewing/Beatmaps/Note.cs]: Not really. LS2OVR C# [reference-implementation](https://github.com/MikuAuahDark/LS2OVR) has similar [struct fields](https://github.com/MikuAuahDark/LS2OVR/blob/7b83d63/LS2OVR/BeatmapTimingMap.cs) to represent "Note". In [this code](https://github.com/MikuAuahDark/LS2OVR/blob/7b83d63/LS2OVR.Unpack/Program.cs#L224-L280), you can see how to convert the internal struct fields back to JSON representation.

In short, it's just hint to the JSON (de)serializer. Not something to worry.

File: Lovewing/Beatmaps/MapDifficulties.cs (hunk 20 above)
Comment (sr229) [reply Lovewing/Beatmaps/MapDifficulties.cs]: Seems to be out of scope for RS so we'll leave that be for now.

File: Lovewing/Beatmaps/RS/RSFormat.cs (hunk 2 above)
Comment (aspriddell) [reply Lovewing/Beatmaps/RS/RSFormat.cs]: @sr229 the ranks in the RS files, so that was what I was going off when writing this. It's fine to remove this if you wish to have a different scoring system to the SIF mobile game and want to do it in a similar fashion to osu! (mainly osu!Taiko)
